### ISSUE TITLE
Changes to fragment array are not present in the owner record's changedAttributes

### ISSUE THREAD

**Issue by nhemanth007** [lytics/ember-data-model-fragments#330]
`changedAttributes` is not returning changes in fragment array on the model.

Ember version used 3.8
Addon version used is v4.0.0

I forked this addon(v4.0.0) and added a unit test case to verify the issue. The test case is not complete, I asserted owner record's changedAttributes with empty string to know what value is returned.

Value returned by changedAttributes is not appropriate. Old property is a fragment object. But new property is a simple object.

 You can find the corresponding test case in https://github.com/nhemanth007/ember-data-model-fragments/commit/44ad92d14c92ca63cd528f4f08fbb6295d2837f2

`names` is a fragment array on person model. String version of data returned by `person.changedAttributes().names` is `[{"type":"name","options":{},"name":"names","owner":{"title":null,"nickName":null,"name":null,"names":[{"first":"Cat","last":"Stark"}],"addresses":[],"titles":[],"hobbies":null,"houses":[],"children":[],"strings":[],"numbers":[],"booleans":[]},"_objectsDirtyIndex":-1,"_objects":[{"first":"Cat","last":"Stark"}],"_lengthDirty":false,"_length":1,"_arrangedContent":[{"first":"Cat","last":"Stark"}],"_originalState":[{"first":"Cat","last":"Stark"}]},{"first":"Cat","last":"Stark"}]`

**Comment by nhemanth007** [comment]
Names on person are changed from `names: [{ first: 'Catelyn', last: 'Stark' }]` to `names: [{ first: 'Cat', last: 'Stark' }]`

**Comment by mattgraham1995** [comment]
I'm seeing a similar issue with my fragmentArrays. I make a change to one of the fragments in the fragmentArray and the parent model will return `true` for `hasDirtyAttributes`, but `changedAttributes()` returns an empty object.

**Comment by VincentMolinie** [comment]
Try with the new beta version I think it might fix your issues 😉

**Comment by VincentMolinie** [comment]
I actually do have the same issue than @mattgraham1995 on version `5.0.0-beta.0`

**Comment by mpirio** [comment]
- ember-source: 3.12.2
- ember-data: 3.12.6
- ember-data-model-fragments: v4.0.0

I have the same "bug" : make a change in a fragmentArray (add or remove an item), `model.hasDirtyAttributes` is `true` but `changedAttributes()` is empty.

I find [`fragmentDidDirty`](https://github.com/lytics/ember-data-model-fragments/blob/v4.0.0/addon/states.js#L166-L173) and [this line in it](https://github.com/lytics/ember-data-model-fragments/blob/v4.0.0/addon/states.js#L170) :

```javascript
record._internalModel._recordData.setDirtyAttribute(key, fragment);
````

But in [`setDirtyAttribute`](https://github.com/emberjs/data/blob/v3.12.6/packages/store/addon/-private/system/model/record-data.ts#L398-L412), ember-data [set `_attributes[key]` line 401](https://github.com/emberjs/data/blob/v3.12.6/packages/store/addon/-private/system/model/record-data.ts#L401) with my fragmentArray value, [set `originalValue`](https://github.com/emberjs/data/blob/v3.12.6/packages/store/addon/-private/system/model/record-data.ts#L406) with the same fragmentArray value and finally compare `_attributes[key]` and `originalValue` so [unset `_attributes[key]`](https://github.com/emberjs/data/blob/v3.12.6/packages/store/addon/-private/system/model/record-data.ts#L410). So `changedAttributes()` will always be empty!

In `states.js` history, I find [this change](https://github.com/lytics/ember-data-model-fragments/commit/e911cb302dcc0f54bd14ef10f579a208c0a52095#diff-1a5e11266321caa561c1642750ad831ed4818a35532c24fda400f185eb478474L170-R170):

```javascript
-    record._internalModel._recordData._attributes[key] = fragment;
-
+    record._internalModel._recordData.setDirtyAttribute(key, fragment);
```

If I revert this change, the bug is resolved... I think the ember-data `recordData.setDirtyAttribute` is not adapted for fragment.

what do you think about it. Perhaps should we revert this change?

**Comment by mpirio** [comment]
An Ember Twiddle to see the bug: https://ember-twiddle.com/93d84d0a212054ff2a0fc146b27917ae

**Comment by mpirio** [comment]
I have a [fix-issue-330 branch in my repo](https://github.com/mpirio/ember-data-model-fragments/tree/fix-issue-330) but I can not PR to this repo because I need a "release-4.0.0" branch here to PR into it. Thanks.

**Comment by stfnio** [comment]
@jakesjews by any chance, do you know why the "release-4.0.0" branch has disappeared? I need the fix discussed above to be merged, as well

**Comment by mpirio** [comment]
No news ?

I try a new call :) @jakesjews @araddon @cohitre @markhayden. Is it possible to create a "release-4.0.0" branch to PR a [fix about this issue](https://github.com/mpirio/ember-data-model-fragments/tree/fix-issue-330)?

Thanks.

**Comment by jakesjews** [comment]
@mpirio were currently in progress moving this repository to ember-adopted-addons https://github.com/lytics/ember-data-model-fragments/issues/382

**Comment by jakesjews** [comment]
If you need the changes right now your best bet is making a fork and branch from commit f1c7060fb7560aaea26760e0270234823b9e20db

**Comment by mpirio** [comment]
Hello @jakesjews.
Thanks for your message.

> @mpirio were currently in progress moving this repository to ember-adopted-addons #382

Yes I read about adopting this addon. It's a good news. But does the procedure block bugfixes on this repository?

> If you need the changes right now your best bet is making a fork and branch from commit [f1c7060](https://github.com/lytics/ember-data-model-fragments/commit/f1c7060fb7560aaea26760e0270234823b9e20db)

That's what I did :). But wouldn't it be better to centralize bugfixes here?

Regards,

**Comment by jakesjews** [comment]
Yes it would but there is some more to the problem [as well](https://github.com/lytics/ember-data-model-fragments/pull/385#issuecomment-724045236) I don't really have much time to work through all the stuff hence moving the repo to adopted-addons

**Comment by mpirio** [comment]
Hello,

It seems the migration to adopted has been made. Really good 👍 . Maybe can I give you my fix for this issue? Would it be possible to create a release-4.0.0 branch so that I can push a PR? Thank you.

**Comment by knownasilya** [comment]
You can submit a PR against the master branch, no need for another branch.

**Comment by mpirio** [comment]
Hello,

Thank you for you message. This PR is for v4.0.0 not 5.0. See https://github.com/mpirio/ember-data-model-fragments/tree/fix-issue-330

**Comment by knownasilya** [comment]
https://github.com/adopted-ember-addons/ember-data-model-fragments/tree/release-v4

**Comment by RobbieTheWagner** [comment]
A lot has changed in this addon. I am going to assume this is no longer a problem and close the issue. If anyone is still having issues and would like to create a reproduction on the latest version, please feel free to do so and we will be glad to try to help!
